File: src/main/java/io/github/skippyall/minions/gui/input/TextInput.java
package io.github.skippyall.minions.gui.input;

import eu.pb4.sgui.api.elements.GuiElementBuilder;
import eu.pb4.sgui.api.gui.AnvilInputGui;
import io.github.skippyall.minions.gui.MinionsGui;
import net.minecraft.network.chat.Component;
import net.minecraft.world.inventory.AnvilMenu;
import net.minecraft.world.item.Items;
import org.jspecify.annotations.Nullable;

import java.util.concurrent.CompletableFuture;
import java.util.function.Function;

public class TextInput<T> extends MinionsGui {
    private final GuiElementBuilder valid = new GuiElementBuilder()
        .setItem(Items.EMERALD_BLOCK)
        .setName(Component.literal("OK"))
        .setCallback(this::onConfirm);

    private final GuiElementBuilder invalid = new GuiElementBuilder()
        .setItem(Items.REDSTONE_BLOCK);

    private final Component title;
    private final String defaultValue;
    private final Function<String, CompletableFuture<Result<T, Component>>> parser;

    private AnvilInputGui gui;

    private @Nullable Result<T, Component> result = null;
    private CompletableFuture<@Nullable T> future = new CompletableFuture<>();

    public TextInput(
            MinionsGui parent,
            Component title,
            String defaultValue,
            Function<String, CompletableFuture<Result<T, Component>>> parser
    ) {
        this.title = title;
        this.defaultValue = defaultValue;
        this.parser = parser;

        super(parent);
        open();
    }

    @Override
    public void open() {
        gui = new AnvilInputGui(viewer, false) {
            @Override
            public void onInput(String input) {
                updateConfirmButton(input);
            }

            @Override
            public void onPlayerClose(boolean success) {
                onBackingClosed();
                if (!future.isDone()) {
                    future.complete(null);
                }
            }
        };

        gui.setTitle(title);
        gui.setDefaultInputValue(defaultValue);
        updateConfirmButton(defaultValue);
        gui.open();
    }

    @Override
    public void closeBacking() {
        gui.close();
    }

    public void updateConfirmButton(String input) {
        parser.apply(input).thenAccept(result -> {
            this.result = result;
            if (result.isSuccess()) {
                gui.setSlot(AnvilMenu.RESULT_SLOT, valid);
            } else {
                Component text = result.getErrorOrThrow();
                gui.setSlot(AnvilMenu.RESULT_SLOT, invalid.setName(text));
            }
        });
    }

    public void onConfirm() {
        if(result != null && result instanceof Result.Success<T, Component> success) {
            future.complete(success.result());
            goBack();
        }
    }

    public static <T> CompletableFuture<@Nullable T> input(
            MinionsGui gui,
            Component title,
            String defaultValue,
            Function<String, CompletableFuture<Result<T, Component>>> parser
    ) {
        TextInput<T> input = new TextInput<>(
                gui,
                title,
                defaultValue,
                parser
        );

        return input.future;
    }

    public static CompletableFuture<@Nullable String> inputString(
            MinionsGui gui,
            Component title,
            String defaultValue
    ) {
        return input(
                gui,
                title,
                defaultValue,
                result -> CompletableFuture.completedFuture(new Result.Success<>(result))
        );
    }

    public static CompletableFuture<@Nullable Long> inputLong(
            MinionsGui gui,
            Component title,
            long defaultValue
    ) {
        return input(
                gui,
                title,
                Long.toString(defaultValue),
                string -> CompletableFuture.completedFuture(Result.wrapCustomError(
                        () -> Long.parseLong(string),
                        Component.translatable("value_type.minions.long.not_long")
                ))
        );
    }

    public static CompletableFuture<@Nullable Double> inputDouble(
            MinionsGui gui,
            Component title,
            double defaultValue
    ) {
        return input(
                gui,
                title,
                Double.toString(defaultValue),
                string -> CompletableFuture.completedFuture(Result.wrapCustomError(
                        () -> Double.parseDouble(string),
                        Component.translatable("value_type.minions.double.not_double")
                ))
        );
    }
}

File: src/main/java/io/github/skippyall/minions/gui/input/TextInput.kt
package io.github.skippyall.minions.gui.input

import eu.pb4.sgui.api.elements.GuiElementBuilder
import eu.pb4.sgui.api.gui.AnvilInputGui
import io.github.skippyall.minions.gui.MinionsGui
import kotlinx.coroutines.launch
import net.minecraft.network.chat.Component
import net.minecraft.world.inventory.AnvilMenu
import net.minecraft.world.item.Items
import java.util.concurrent.CompletableFuture

class TextInput<T>(
    parent: MinionsGui,
    val title: Component,
    val defaultValue: String,
    val parser: suspend (String) -> Result<T, Component>
) : MinionsGui(parent) {
    private val valid: GuiElementBuilder = GuiElementBuilder()
        .setItem(Items.EMERALD_BLOCK)
        .setName(Component.literal("OK"))
        .setCallback(Runnable { this.onConfirm() })

    private val invalid: GuiElementBuilder = GuiElementBuilder()
        .setItem(Items.REDSTONE_BLOCK)

    private lateinit var gui: AnvilInputGui

    private var result: Result<T, Component>? = null
    val future = CompletableFuture<T?>()

    init {
        open()
    }

    override fun open() {
        gui = object : AnvilInputGui(viewer, false) {
            override fun onInput(input: String) {
                updateConfirmButton(input)
            }

            override fun onPlayerClose(success: Boolean) {
                onBackingClosed()
                if (!future.isDone) {
                    future.complete(null)
                }
            }
        }

        gui.setTitle(title)
        gui.setDefaultInputValue(defaultValue)
        updateConfirmButton(defaultValue)
        gui.open()
    }

    override fun closeBacking() {
        gui.close()
    }

    fun updateConfirmButton(input: String) {
        scope.launch {
            val result = parser(input)
            this@TextInput.result = result
            if (result.isSuccess()) {
                gui.setSlot(AnvilMenu.RESULT_SLOT, valid)
            } else {
                val text = result.getErrorOrThrow()
                gui.setSlot(AnvilMenu.RESULT_SLOT, invalid.setName(text))
            }
        }
    }

    fun onConfirm() {
        result?.ifSuccess { success: T ->
            future.complete(success)
            goBack()
        }
    }

    companion object {
        @JvmStatic
        fun <T>input(
            gui: MinionsGui,
            title: Component,
            defaultValue: String,
            parser: suspend (String) -> Result<T, Component>,
        ): CompletableFuture<T?> {
            val input = TextInput(
                parent = gui,
                title = title,
                defaultValue = defaultValue,
                parser = parser,
            )

            return input.future
        }

        @JvmStatic
        fun inputString(
            gui: MinionsGui,
            title: Component,
            defaultValue: String,
        ): CompletableFuture<String?> {
            return input<String>(
                gui = gui,
                title = title,
                defaultValue = defaultValue,
                parser = { result: String? -> Result.Success<String, Component>(result) },
            )
        }

        @JvmStatic
        fun inputLong(
            gui: MinionsGui,
            title: Component,
            defaultValue: Long,
        ): CompletableFuture<Long?> {
            return input<Long>(
                gui = gui,
                title = title,
                defaultValue = defaultValue.toString(),
                parser = { string ->
                    Result.wrapCustomError<Long, Component>(
                        { string.toLong() },
                        Component.translatable("minions.command.input.int.fail")
                    )
                },
            )
        }

        @JvmStatic
        fun inputDouble(
            gui: MinionsGui,
            title: Component,
            defaultValue: Double,
        ): CompletableFuture<Double?> {
            return input<Double>(
                gui = gui,
                title = title,
                defaultValue = defaultValue.toString(),
                parser = { string ->
                    Result.wrapCustomError<Double, Component>(
                        { string.toDouble() },
                        Component.translatable("minions.command.input.int.fail")
                    )
                },
            )
        }
    }
}
